### ISSUE TITLE
Deepgram module latency issue

### ISSUE THREAD

**Issue by kapilrupavatiya** [jambonz/freeswitch-modules#18]
Hi First off, thank you for building and maintaining this amazing package. Our company is delivering real time voice capabilities to millions of customers of large enterprises in India, and want to leverage the Deepgram module in production setting.
We use Lua scripting language. We've managed to integrate the module with Freeswitch setup successfully. However, we're noticing a latency of at least 3 seconds between when the user stops speaking and when Freeswitch receives the response from Deepgram servers. Additionally, in the initial phase of the call, we're seeing a number of "discarding empty deepgram transcript" messages on the console. I have attached logs below. The speaker started speaking at 20:31:40 and spoke for approx. 2 seconds. The server sent the transcription around 20:31:46. This means there is a latency of 4 seconds. We would like to reduce the latency well below 1 second. Could you please help us with this.
```
2024-03-07 20:31:40.443757 99.93% [INFO] mod_deepgram_transcribe.c:154 start transcribing hi interim
2024-03-07 20:31:40.443757 99.93% [DEBUG] dg_transcribe_glue.cpp:336 path: /v1/listen?tier=enhanced&model=general&language=hi&smart_format=true&no_delay=true&punctuate=true&interim_results=true&encoding=linear16&sample_rate=8000
2024-03-07 20:31:40.443757 99.93% [DEBUG] dg_transcribe_glue.cpp:377 (64) no resampling needed for this call
2024-03-07 20:31:40.443757 99.93% [DEBUG] dg_transcribe_glue.cpp:380 (64) fork_data_init
2024-03-07 20:31:40.443757 99.93% [DEBUG] dg_transcribe_glue.cpp:452 connecting now
2024-03-07 20:31:40.443757 99.93% [DEBUG] dg_transcribe_glue.cpp:454 connection in progress
2024-03-07 20:31:40.443757 99.93% [DEBUG] mod_deepgram_transcribe.c:41 Got SWITCH_ABC_TYPE_INIT.
2024-03-07 20:31:40.443757 99.93% [DEBUG] switch_core_media_bug.c:976 Attaching BUG to sofia/tata-profile/09429515176
2024-03-07 20:31:40.443757 99.93% [DEBUG] mod_deepgram_transcribe.c:98 added media bug for dg transcribe
2024-03-07 20:31:40.443757 99.93% [DEBUG] switch_cpp.cpp:86 bound to CUSTOM deepgram_transcribe::transcription
2024-03-07 20:31:40.463756 99.93% [DEBUG] switch_core_io.c:448 Setting BUG Codec PCMU:0
2024-03-07 20:31:41.783759 99.93% [INFO] dg_transcribe_glue.cpp:279 connection successful
2024-03-07 20:31:43.203760 99.93% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-07 20:31:44.103758 99.93% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-07 20:31:44.303757 99.93% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-07 20:31:45.283757 99.93% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-07 20:31:46.463765 99.90% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":2.0189373,"start":2.26,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"हां बोलो","confidence":0.9941406,"words":[{"word":"हां","start":2.26,"end":2.76,"confidence":0.84228516,"punctuated_word":"हां"},{"word":"बोलो","start":3.2694688,"end":3.7694688,"confidence":0.9941406,"punctuated_word":"बोलो"}]}]},"metadata":{"request_id":"b0a43754-4336-4821-81ec-2eeda6d5bb4d","model_info":{"name":"general-enhanced","version":"1982-11-21.24709","arch":"polaris"},"model_uuid":"83d5f3ce-9a16-4eb9-affe-83bd9ad10f69"}}
2024-03-07 20:31:47.443757 99.90% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":3.0189373,"start":2.26,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"हां बोलो क्या काम है?","confidence":0.89453125,"words":[{"word":"हां","start":3.7792401,"end":3.97914,"confidence":0.78222656,"punctuated_word":"हां"},{"word":"बोलो","start":3.97914,"end":4.33896,"confidence":0.9980469,"punctuated_word":"बोलो"},{"word":"क्या","start":4.33896,"end":4.5388603,"confidence":0.8520508,"punctuated_word":"क्या"},{"word":"काम","start":4.5388603,"end":4.81872,"confidence":0.9995117,"punctuated_word":"काम"},{"word":"है","start":4.81872,"end":5.2789373,"confidence":0.89453125,"punctuated_word":"है?"}]}]},"metadata":{"request_id":"b0a43754-4336-4821-81ec-2eeda6d5bb4d","model_info":{"name":"general-enhanced","version":"1982-11-21.24709","arch":"polaris"},"model_uuid":"83d5f3ce-9a16-4eb9-affe-83bd9ad10f69"}}
2024-03-07 20:31:47.623757 99.90% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":3.2500002,"start":2.26,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"हां बोलो क्या काम है?","confidence":0.9238281,"words":[{"word":"हां","start":3.7792401,"end":3.97914,"confidence":0.78222656,"punctuated_word":"हां"},{"word":"बोलो","start":3.97914,"end":4.33896,"confidence":0.99853516,"punctuated_word":"बोलो"},{"word":"क्या","start":4.33896,"end":4.5388603,"confidence":0.8540039,"punctuated_word":"क्या"},{"word":"काम","start":4.5388603,"end":4.81872,"confidence":0.9995117,"punctuated_word":"काम"},{"word":"है","start":4.81872,"end":5.31872,"confidence":0.9238281,"punctuated_word":"है?"}]}]},"metadata":{"request_id":"b0a43754-4336-4821-81ec-2eeda6d5bb4d","model_info":{"name":"general-enhanced","version":"1982-11-21.24709","arch":"polaris"},"model_uuid":"83d5f3ce-9a16-4eb9-affe-83bd9ad10f69"}}
```

**Comment by davehorton** [comment]
`tier` has been deprecated by Deepgram and you should not use it.  For language 'hi' I suggest trying a different model - nova-2

**Comment by kapilrupavatiya** [comment]
Thanks, @davehorton for the quick revert. We changed the configuration based on your recommendation. Still, we are facing the same issue. Please see the below log:

```
2024-03-09 01:41:11.572642 99.90% [INFO] mod_deepgram_transcribe.c:154 start transcribing hi interim
2024-03-09 01:41:11.572642 99.90% [DEBUG] dg_transcribe_glue.cpp:336 path: /v1/listen?&model=nova-2&language=hi&interim_results=true&endpointing=500&encoding=linear16&sample_rate=8000
2024-03-09 01:41:11.572642 99.90% [DEBUG] dg_transcribe_glue.cpp:377 (17) no resampling needed for this call
2024-03-09 01:41:11.572642 99.90% [DEBUG] dg_transcribe_glue.cpp:380 (17) fork_data_init
2024-03-09 01:41:11.572642 99.90% [DEBUG] dg_transcribe_glue.cpp:452 connecting now
2024-03-09 01:41:11.572642 99.90% [DEBUG] dg_transcribe_glue.cpp:454 connection in progress
2024-03-09 01:41:11.572642 99.90% [DEBUG] mod_deepgram_transcribe.c:41 Got SWITCH_ABC_TYPE_INIT.
2024-03-09 01:41:11.572642 99.90% [DEBUG] switch_core_media_bug.c:976 Attaching BUG to sofia/tata-profile/09429515176
2024-03-09 01:41:11.572642 99.90% [DEBUG] mod_deepgram_transcribe.c:98 added media bug for dg transcribe
2024-03-09 01:41:11.572642 99.90% [DEBUG] switch_cpp.cpp:86 bound to CUSTOM deepgram_transcribe::transcription
2024-03-09 01:41:11.592641 99.90% [DEBUG] switch_core_io.c:448 Setting BUG Codec PCMU:0
2024-03-09 01:41:12.572640 99.90% [INFO] dg_transcribe_glue.cpp:279 connection successful
2024-03-09 01:41:13.952640 99.87% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 01:41:15.032643 99.87% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 01:41:15.092640 99.87% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 01:41:16.232652 99.83% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 01:41:17.332640 99.83% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 01:41:17.812640 99.80% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 01:41:18.752640 99.80% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":2.5889373,"start":3.23,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"मेरा नाम कपिल है","confidence":0.9941406,"words":[{"word":"मेरा","start":4.67,"end":4.99,"confidence":0.9941406},{"word":"नाम","start":4.99,"end":5.23,"confidence":1.0},{"word":"कपिल","start":5.23,"end":5.63,"confidence":0.9819336},{"word":"है","start":5.63,"end":5.8189373,"confidence":0.5703125}]}]},"metadata":{"request_id":"5dabdd14-25c1-4f25-a273-8261ab9b42fb","model_info":{"name":"2-general-nova","version":"2023-11-21.25539","arch":"nova-2"},"model_uuid":"6cc5d67a-4cb5-413c-9336-519017c4f856"}}
2024-03-09 01:41:19.572640 99.80% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":3.3289375,"start":3.23,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"मेरा नाम कपिल है","confidence":0.99316406,"words":[{"word":"मेरा","start":4.67,"end":4.99,"confidence":0.99316406},{"word":"नाम","start":4.99,"end":5.23,"confidence":1.0},{"word":"कपिल","start":5.23,"end":5.71,"confidence":0.9873047},{"word":"है","start":5.71,"end":6.21,"confidence":0.9658203}]}]},"metadata":{"request_id":"5dabdd14-25c1-4f25-a273-8261ab9b42fb","model_info":{"name":"2-general-nova","version":"2023-11-21.25539","arch":"nova-2"},"model_uuid":"6cc5d67a-4cb5-413c-9336-519017c4f856"}}
2024-03-09 01:41:19.772640 99.80% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":3.5089374,"start":3.23,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"मेरा नाम कपिल है","confidence":0.99316406,"words":[{"word":"मेरा","start":4.67,"end":4.99,"confidence":0.99316406},{"word":"नाम","start":4.99,"end":5.23,"confidence":1.0},{"word":"कपिल","start":5.23,"end":5.71,"confidence":0.98828125},{"word":"है","start":5.71,"end":6.21,"confidence":0.9770508}]}]},"metadata":{"request_id":"5dabdd14-25c1-4f25-a273-8261ab9b42fb","model_info":{"name":"2-general-nova","version":"2023-11-21.25539","arch":"nova-2"},"model_uuid":"6cc5d67a-4cb5-413c-9336-519017c4f856"}}
2024-03-09 01:41:20.072639 99.80% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":3.8889375,"start":3.23,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"मेरा नाम कपिल है","confidence":0.99316406,"words":[{"word":"मेरा","start":4.67,"end":4.99,"confidence":0.99316406},{"word":"नाम","start":4.99,"end":5.23,"confidence":1.0},{"word":"कपिल","start":5.23,"end":5.71,"confidence":0.98828125},{"word":"है","start":5.71,"end":6.21,"confidence":0.98291016}]}]},"metadata":{"request_id":"5dabdd14-25c1-4f25-a273-8261ab9b42fb","model_info":{"name":"2-general-nova","version":"2023-11-21.25539","arch":"nova-2"},"model_uuid":"6cc5d67a-4cb5-413c-9336-519017c4f856"}}
2024-03-09 01:41:20.412640 99.80% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":4.2289376,"start":3.23,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"मेरा नाम कपिल है","confidence":0.99316406,"words":[{"word":"मेरा","start":4.67,"end":4.99,"confidence":0.99316406},{"word":"नाम","start":4.99,"end":5.23,"confidence":1.0},{"word":"कपिल","start":5.23,"end":5.71,"confidence":0.98828125},{"word":"है","start":5.71,"end":6.21,"confidence":0.9873047}]}]},"metadata":{"request_id":"5dabdd14-25c1-4f25-a273-8261ab9b42fb","model_info":{"name":"2-general-nova","version":"2023-11-21.25539","arch":"nova-2"},"model_uuid":"6cc5d67a-4cb5-413c-9336-519017c4f856"}}
2024-03-09 01:41:21.012640 99.80% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":4.7900004,"start":3.23,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"मेरा नाम कपिल है","confidence":0.99316406,"words":[{"word":"मेरा","start":4.67,"end":4.99,"confidence":0.99316406},{"word":"नाम","start":4.99,"end":5.23,"confidence":1.0},{"word":"कपिल","start":5.23,"end":5.71,"confidence":0.9873047},{"word":"है","start":5.71,"end":6.21,"confidence":0.9892578}]}]},"metadata":{"request_id":"5dabdd14-25c1-4f25-a273-8261ab9b42fb","model_info":{"name":"2-general-nova","version":"2023-11-21.25539","arch":"nova-2"},"model_uuid":"6cc5d67a-4cb5-413c-9336-519017c4f856"}}
``` 

In this, the User started speaking at ~2024-03-09 01:41:12 and stopped speaking ~2024-03-09 01:41:16. But if you see from deepgram till ~2024-03-09 01:41:18 we are getting `discarding empty deepgram transcript` and then after we are getting actual transcribe in the result.

Also, The strange thing is, If we see the final result event as below:
```
2024-03-09 01:41:21.012640 99.80% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":4.7900004,"start":3.23,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"मेरा नाम कपिल है","confidence":0.99316406,"words":[{"word":"मेरा","start":4.67,"end":4.99,"confidence":0.99316406},{"word":"नाम","start":4.99,"end":5.23,"confidence":1.0},{"word":"कपिल","start":5.23,"end":5.71,"confidence":0.9873047},{"word":"है","start":5.71,"end":6.21,"confidence":0.9892578}]}]},"metadata":{"request_id":"5dabdd14-25c1-4f25-a273-8261ab9b42fb","model_info":{"name":"2-general-nova","version":"2023-11-21.25539","arch":"nova-2"},"model_uuid":"6cc5d67a-4cb5-413c-9336-519017c4f856"}}
``` 

It says the duration is 4.79 seconds which is correct, But if you see the start it says 3.23, Which I feel is not correct, Because the user started speaking as soon as the transcribe started. But somehow due to buffer or some other configuration in starting 3.23 seconds, deepgram gets empty audio and due to that only it returns `discarding empty deepgram transcript` till 2024-03-09 01:41:18.

Let me know if you needed any other information from my side.

**Comment by davehorton** [comment]
at this point you should be troubleshooting this with deepgram.  They are providing the recognition service, and the 4.79 seconds that you cite is merely us logging what they are sending us.  Have you considered that maybe deepgram just isnt that good at transcribing Hindi?  In my experience their english transcription is lightning fast and good, other languages...not so much.

But I would suggest trying to open a ticket with them

**Comment by kapilrupavatiya** [comment]
@davehorton when we see this latency first time, we immediately tested the deepgram streaming service using a standalone python script to make sure latency is from deepgram or not. But in that python script we got a response immediately within milliseconds, so there is no issue from deepgram.

**Comment by kapilrupavatiya** [comment]
@davehorton Based on your suggestion I also tested with english language, But it also issue is same. Please see below logs for the same: 

```
2024-03-09 10:03:23.532641 99.90% [INFO] mod_deepgram_transcribe.c:154 start transcribing en-US interim
2024-03-09 10:03:23.532641 99.90% [DEBUG] dg_transcribe_glue.cpp:336 path: /v1/listen?&model=nova-2&language=en-US&interim_results=true&endpointing=500&encoding=linear16&sample_rate=8000
2024-03-09 10:03:23.532641 99.90% [DEBUG] dg_transcribe_glue.cpp:377 (18) no resampling needed for this call
2024-03-09 10:03:23.532641 99.90% [DEBUG] dg_transcribe_glue.cpp:380 (18) fork_data_init
2024-03-09 10:03:23.532641 99.90% [DEBUG] dg_transcribe_glue.cpp:452 connecting now
2024-03-09 10:03:23.532641 99.90% [DEBUG] dg_transcribe_glue.cpp:454 connection in progress
2024-03-09 10:03:23.532641 99.90% [DEBUG] mod_deepgram_transcribe.c:41 Got SWITCH_ABC_TYPE_INIT.
2024-03-09 10:03:23.532641 99.90% [DEBUG] switch_core_media_bug.c:976 Attaching BUG to sofia/tata-profile/09429515176
2024-03-09 10:03:23.532641 99.90% [DEBUG] mod_deepgram_transcribe.c:98 added media bug for dg transcribe
2024-03-09 10:03:23.532641 99.90% [DEBUG] switch_cpp.cpp:86 bound to CUSTOM deepgram_transcribe::transcription
2024-03-09 10:03:23.552641 99.90% [DEBUG] switch_core_io.c:448 Setting BUG Codec PCMU:0
2024-03-09 10:03:24.652640 99.90% [INFO] dg_transcribe_glue.cpp:279 connection successful
2024-03-09 10:03:25.912642 99.90% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 10:03:26.872640 99.90% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 10:03:27.372640 99.90% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 10:03:28.372640 99.93% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 10:03:28.672640 99.97% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 10:03:29.092640 99.97% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 10:03:29.732640 99.97% [DEBUG] dg_transcribe_glue.cpp:305 discarding empty deepgram transcript
2024-03-09 10:03:30.712640 99.97% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":3.3189373,"start":2.5,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"hello my name is","confidence":1.0,"words":[{"word":"hello","start":4.66,"end":5.14,"confidence":0.9951172},{"word":"my","start":5.14,"end":5.38,"confidence":1.0},{"word":"name","start":5.38,"end":5.62,"confidence":1.0},{"word":"is","start":5.62,"end":5.8189373,"confidence":0.9970703}]}]},"metadata":{"request_id":"d9d78e04-3126-4632-b392-7a78336fca69","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"}}
2024-03-09 10:03:30.752640 99.97% [ERR] switch_cpp.cpp:1465 loop.lua	schema	Kq4oP1t=XnbCGO0VtLi_zECJv3DyLiiizaxb2Q	123445	START INNER LOOP
2024-03-09 10:03:32.032640 99.97% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":4.64,"start":2.5,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"hello my name is kapil","confidence":1.0,"words":[{"word":"hello","start":4.74,"end":5.2200003,"confidence":0.99365234},{"word":"my","start":5.2200003,"end":5.46,"confidence":1.0},{"word":"name","start":5.46,"end":5.7,"confidence":1.0},{"word":"is","start":5.7,"end":5.9399996,"confidence":1.0},{"word":"kapil","start":5.9399996,"end":6.4399996,"confidence":0.73250324}]}]},"metadata":{"request_id":"d9d78e04-3126-4632-b392-7a78336fca69","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"}}
``` 

In this, the User started speaking at ~2024-03-09 10:03:24 and stopped speaking ~2024-03-09 10:03:28. But if you see from deepgram till ~2024-03-09 10:03:32 we are getting `discarding empty deepgram transcript` and then after we are getting actual transcribe in the result. Also on below result response, We can see it started at 2.5 second:
```
2024-03-09 10:03:32.032640 99.97% [DEBUG] dg_transcribe_glue.cpp:309 deepgram message: {"type":"Results","channel_index":[0,1],"duration":4.64,"start":2.5,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"hello my name is kapil","confidence":1.0,"words":[{"word":"hello","start":4.74,"end":5.2200003,"confidence":0.99365234},{"word":"my","start":5.2200003,"end":5.46,"confidence":1.0},{"word":"name","start":5.46,"end":5.7,"confidence":1.0},{"word":"is","start":5.7,"end":5.9399996,"confidence":1.0},{"word":"kapil","start":5.9399996,"end":6.4399996,"confidence":0.73250324}]}]},"metadata":{"request_id":"d9d78e04-3126-4632-b392-7a78336fca69","model_info":{"name":"2-general-nova","version":"2024-01-11.36317","arch":"nova-2"},"model_uuid":"1dbdfb4d-85b2-4659-9831-16b3c76229aa"}}
``` 

So we suspect that somehow from FreeSWITCh it streaming blank audio for starting/beginning 3-4 seconds and then start actual user audio.

**Comment by davehorton** [comment]
The delays that you experience for English I have not seen and I am pretty sure they are not caused by this module.  It would be useful if you wanted to test against my system to see if you have the same or different results.  You can create a free account at https://jambonz.cloud and use the echo test app that is provided after adding a sip trunk and your deepgram api key

**Comment by neilp9** [comment]
Hi Dave, we also suspect there's a build or configuration issue causing this delay. We need this working in our own infrastructure. Could you help us troubleshoot our build, config, and app? We've sent you an email to engage. We expect it would not take you long to spot a shallow issue.

**Comment by neilp9** [comment]
We have gotten a handle of the issue, it seems to be related to how we set `ignore_early_media` in the call origination. The behavior we seem to be seeing is that when it's set to `true`, FS buffers all the early media and sends it through the socket first before live call speech. Is this behavior you've seen, and/or is expected?

When we set it to `false`, the media doesn't buffer and we immediately start getting non-blank transcripts from DG. This requires us adding logic in our app to initiate our agent after call has been picked up.

We'd like to not have to bring call flow management into our app, which is unreliable. Is there a way to "flush" early media from being sent through the socket once call picks up/becomes active?

**Comment by neilp9** [comment]
@davehorton we've isolated another issue causing the websocket pipe to be disturbed resulting in delayed transcription: in our Lua app's main turn-taking loop, we make an api request to generate our agent's response after each caller final utterance. This API request is done using [luasec](https://github.com/lunarmodules/luasec). We're seeing that after every API call (HTTPS request), the transctiption delay widens by a couple seconds. It's like the request (which itself returns in milliseconds) is causing the socket connection to fall further and further behind.

 ```
local json = require("cjson")
local https = require("ssl.https")
local ltn12 = require("ltn12")

  local response_body = {}
    local _, response_code, response_headers= https.request {
      url = url,
      method = "POST",
      headers = headers,
      source = ltn12.source.string(json.encode(body)),
      sink = ltn12.sink.table(response_body)
    }
```

Any ideas on what is causing this and why the DG websocket would be disrupted by a HTTPS request each turn? We tried stopping/starting the DG websocket in between turns, but that didn't fix it.

Would appreciate your suggestions on how we can address this
